File: raw/github/paypal/paypal-android-sdk-demo-app/snapshots/2026-08-16-d1137d5/files/app/src/main/java/com/firstapp/paypaldemo/cardcheckout/CardPaymentViewModel.kt
package com.firstapp.paypaldemo.cardcheckout

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.firstapp.paypaldemo.Constants.CLIENT_ID
import com.firstapp.paypaldemo.service.Amount
import com.firstapp.paypaldemo.service.DemoMerchantAPI
import com.firstapp.paypaldemo.service.PurchaseUnit
import com.paypal.android.cardpayments.Card
import com.paypal.android.cardpayments.CardApproveOrderResult
import com.paypal.android.cardpayments.CardClient
import com.paypal.android.cardpayments.CardRequest
import com.paypal.android.cardpayments.threedsecure.SCA
import com.paypal.android.corepayments.CoreConfig
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import jakarta.inject.Inject
import kotlinx.coroutines.launch

/**
 * A separate ViewModel that:
 *  - Creates an order on your merchant server.
 *  - Approves that order with the PayPal card client (callback-based).
 *  - Captures the order on your server.
 *
 * No 3DS logic here: If a flow requires 3DS, it won't be handled.
 */
@HiltViewModel
class CardPaymentViewModel @Inject constructor(
    @ApplicationContext context: Context
) : ViewModel() {

    private val config = CoreConfig(CLIENT_ID)
    private val cardClient = CardClient(context, config)

    var isLoading: Boolean = false
        private set
    var errorMessage: String = ""
        private set

    /**
     * The main “checkoutWithCard” function, using the validated Card details.
     */
    fun checkoutWithCard(
        card: Card,
        amount: Double,
        onSuccess: (String) -> Unit,
        onFailure: (String) -> Unit
    ) {
        isLoading = true
        errorMessage = ""

        viewModelScope.launch {
            try {
                val order = DemoMerchantAPI.createOrder(
                    intent = "CAPTURE",
                    purchaseUnits = listOf(
                        PurchaseUnit(
                            amount = Amount(currencyCode = "USD", value = amount.toString())
                        )
                    )
                )
                val orderId = order.id
                println("✅ Created order: $orderId, status: ${order.status}")

                val sdkCard = Card(
                    number = card.number,
                    expirationMonth = card.expirationMonth,
                    expirationYear = card.expirationYear,
                    securityCode = card.securityCode
                )
                val request = CardRequest(
                    orderId,
                    sdkCard,
                    returnUrl = "com.firstapp.paypal.demo://example.com",
                    sca = SCA.SCA_WHEN_REQUIRED
                )

                cardClient.approveOrder(request) { approveResult ->
                    when (approveResult) {
                        is CardApproveOrderResult.Success -> {
                            println("✅ Card approved. orderId=${approveResult.orderId}, status=${approveResult.status}")
                            // 3) Capture on your server
                            viewModelScope.launch {
                                try {
                                    val finalOrder =
                                        DemoMerchantAPI.completeOrder(orderId, "CAPTURE")
                                    println("✅ Order captured: ${finalOrder.id}, status=${finalOrder.status}")
                                    isLoading = false
                                    onSuccess(finalOrder.id)
                                } catch (e: Exception) {
                                    isLoading = false
                                    errorMessage = "Capture error: ${e.message}"
                                    onFailure(errorMessage)
                                }
                            }
                        }

                        is CardApproveOrderResult.AuthorizationRequired -> {
                            // 3DS flow required if the user had set sca, but we said “no 3ds,”
                            // so we can treat it as an error or skip handling it:
                            isLoading = false
                            errorMessage = "Auth required but 3DS not supported in this example."
                            onFailure(errorMessage)
                        }

                        is CardApproveOrderResult.Failure -> {
                            isLoading = false
                            errorMessage = "Card approve failed: ${approveResult.error.message}"
                            onFailure(errorMessage)
                        }
                    }
                }

            } catch (createErr: Exception) {
                // If the merchant server createOrder fails
                isLoading = false
                errorMessage = "Create order error: ${createErr.message}"
                onFailure(errorMessage)
            }
        }
    }
}

File: raw/github/paypal/paypal-ios-sdk-demo-app/snapshots/2026-08-16-047a50e/files/PayPalDemo/ViewModels/CardPaymentViewModel.swift
import Foundation
import CardPayments
import CorePayments
import FraudProtection

class CardPaymentViewModel: ObservableObject {

    private var cardClient: CardClient?
    private var payPalDataCollector: PayPalDataCollector?

    func checkoutWith(
        card: Card,
        amount: String,
        intent: Intent,
        sca: SCA
    ) async throws -> Order {
        do {
            async let config = try await DemoMerchantAPI.shared.getCoreConfig()

            let order = try await DemoMerchantAPI.shared.createOrder(
                orderParams: CreateOrderParams(
                    applicationContext: nil,
                    intent: intent.rawValue,
                    purchaseUnits: [PurchaseUnit(amount: Amount(currencyCode: "USD", value: amount))]
                )
            )
            print("✅ Order created with orderID: \(order.id) with status: \(order.status)")

            cardClient = try await CardClient(config: config)
            guard let cardClient = cardClient else {
                throw NSError(domain: "CardClientError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Card client could not be initialized."])
            }

            let cardRequest = CardRequest(orderID: order.id, card: card, sca: sca)
            let cardResult = try await cardClient.approveOrder(request: cardRequest)
            print("✅ Card approval returned with CardResult \norderID: \(cardResult.orderID) \nstatus: \(String(describing: cardResult.status)) \ndidAttemptThreeDSecureAuthentication: \(cardResult.didAttemptThreeDSecureAuthentication)")

            payPalDataCollector = try await PayPalDataCollector(config: config)
            let payPalClientMetadataID = payPalDataCollector?.collectDeviceData()

            let completedOrder = try await DemoMerchantAPI.shared.completeOrder(
                orderID: order.id,
                payPalClientMetadataID: payPalClientMetadataID,
                intent: intent)
            print("✅ Capture returned with orderID: \(completedOrder.id) with status: \(completedOrder.status) ")
            return completedOrder
        } catch let error {
            print("❌ Failed in checkout with card: \(error.localizedDescription)")
            throw error
        }
    }
}
